added Intercepting Apple promotional purchases

Ponyu-dev opened the pull request:
https://docs.unity3d.com/Manual/UnityIAPiOSMAS.html

Ponyu-dev committed:
added Intercepting Apple promotional purchases

IvanMurzak commented on Assets/_PackageRoot/Scripts/UnityIAPInitializer.cs:
```diff
@@ -217,5 +231,9 @@ public		virtual		void															OnPurchaseFailed				(Product i, PurchaseFail
+
+		public		virtual		void															OnPromotionalPurchase			(Product i) => onProductPromotionalPurchase.OnNext(i);
+
+		public		virtual		void															ContinuePromotionalPurchases	()			=> extensionsApple.ContinuePromotionalPurchases();
```
In case the method is executed in **non-iOS platform**, it can produce **NullReferenceException**, because `extensionsApple` could be null.

IvanMurzak commented on Assets/_PackageRoot/Scripts/UnityIAPInitializer.cs:
```diff
@@ -217,5 +231,9 @@ public		virtual		void															OnPurchaseFailed				(Product i, PurchaseFail
 			});
 		}
+
+		public		virtual		void															OnPromotionalPurchase			(Product i) => onProductPromotionalPurchase.OnNext(i);
```
This method should not be accessible outside of the class, better to have this one as `protected`

IvanMurzak replied:
Also would be awesome to have logs in it

IvanMurzak commented on Assets/_PackageRoot/Scripts/StoreSO.cs:
```diff
@@ -301,5 +315,12 @@ public		void			RestorePurchases				()
 			unityIAPInitializer?.RestorePurchases();
 		}
+
+		// You must call IAppleExtensions.ContinuePromotionalPurchases() to continue with the promotional purchase.
```
There is no `IAppleExtensions.ContinuePromotionalPurchases` method for a user

IvanMurzak commented on Assets/_PackageRoot/Scripts/UnityIAPInitializer.cs:
```diff
@@ -159,6 +162,11 @@ public		virtual		void															Init							(IEnumerable<StoreSellable> iapSe
+
+			// On iOS and tvOS we can intercept promotional purchases that come directly from the App Store.
+			// On other platforms this will have no effect; OnPromotionalPurchase will never be called.
+			builder.Configure<IAppleConfiguration>().SetApplePromotionalPurchaseInterceptorCallback(OnPromotionalPurchase);
```
I believe the line should be moved to `UnityIAPInitializer.OnInitialized` method. Because only after `OnInitialized` event the extensions could be taken.

IvanMurzak commented on Assets/_PackageRoot/Scripts/UnityIAPInitializer.cs:
```diff
@@ -180,6 +188,12 @@ public		virtual		void															OnInitialized					(IStoreController controll
+			// Set all these products to be visible in the user's App Store
+			foreach (var item in controller.products.all)
+			{
+				if (item.availableToPurchase) extensionsApple.SetStorePromotionVisibility(item, AppleStorePromotionVisibility.Show);				
```
Please add null check:
`extensionsApple.SetStoreProm...` -> `extensionsApple?.SetStoreProm...`

IvanMurzak commented on Assets/_PackageRoot/Scripts/StoreSO.cs:
```diff
@@ -12,37 +12,40 @@ namespace Project.Store
+												private					Subject<bool>														onRestorePurchasesCompleted		= new Subject<bool>();
+												public					IObservable<bool>													OnRestorePurchasesCompleted		=> onRestorePurchasesCompleted;
+		
+												private					Subject<StoreSellable>												onProductPromotionalPurchase	= new Subject<StoreSellable>();
+												public					IObservable<StoreSellable>											OnProductPromotionalPurchase	=> onProductPromotionalPurchase;
```
Please use past time in events naming `...Purchase` -> `...Purchased`

IvanMurzak commented on Assets/_PackageRoot/Scripts/UnityIAPInitializer.cs:
```diff
@@ -25,7 +25,10 @@ public partial class UnityIAPInitializer : IStoreListener
 		public					IObservable<bool>												OnRestorePurchasesCompleted		=> onRestorePurchasesCompleted == null ? onRestorePurchasesCompleted = new Subject<bool>() : onRestorePurchasesCompleted;
-	
+		
+								Subject<Product>												onProductPromotionalPurchase	= new Subject<Product>();
+		public					IObservable<Product>											OnProductPromotionalPurchase	=> onProductPromotionalPurchase == null ? onProductPromotionalPurchase = new Subject<Product>() : onProductPromotionalPurchase;
```
Please use past time for events naming
`...Purchase` -> `...Purchased`

IvanMurzak reviewed (changes requested):
@Ponyu-dev  Is the `OnProductPurchased` called if user bought product through Apple Promotion? It is important, because only through this event gonna executed `ApplyPurchase`, in other case user won't receive the product in the end.

Ponyu-dev committed:
Fixed remarks

Ponyu-dev committed:
Added selection "AppleStorePromotionVisibility"

Ponyu-dev committed:
Update StoreSO.cs

Sometimes you need to remember that there was an Apple promotional purchases.
To first go to some screen and then continue the purchase.

Ponyu-dev committed:
made medium changes

IvanMurzak commented on Assets/_PackageRoot/Scripts/StoreSO.cs:
```diff
@@ -100,6 +110,7 @@ List<StoreSellable>								SellablesList
 												public					bool											ValidateCurrency		(string currency)   => currency == null ? false : AllCurrenciesNames().Contains(currency);
 												public					bool											ValidateCategory		(string category)	=> category == null ? false : categories.ContainsKey(category);
 
+												public					bool											IsProductApplePromotional					=> _productApplePromotional != null;
```
Looks like we don't really need the property `IsProductApplePromotional` with the variable `_productApplePromotional` at all. Because the promotional is available to receive from related events. If so, need to delete the property + variable.
